### ISSUE TITLE
Performance issues with DialogRPT + DialoGPT

### ISSUE THREAD

**Issue by pablogranolabar** [golsun/DialogRPT#6]
Hi again @golsun,

I've been working with DialogRPT using DialoGPT-large for dialog generation and have hit some performance issues that aren't present when using just DialoGPT-large. Round trip responses using CPU inference are just a few seconds with gpt2-large but whenever DialogRPT is used with the DialoGPT-large checkpoint, performance grinds to a halt. With GPU inference I can run gpt2-large on a 6GB GPU but with DialogRPT I get OOM. I understand that there are multiple models running with the combination of DialogRPT + DialoGPT which is the obvious culprit, is there any way to serialize execution of the two models to prevent these resource consumption issues?

**Comment by golsun** [comment]
hi @pablogranolabar ,

I can think of several potential reasons of OOM:
* `torch.no_grad` which avoids grad taking memory
it was already applied in [scorer](https://github.com/golsun/DialogRPT/blob/e25ef21a7961e2045405fad4249c15b99876e7a8/src/model.py#L50), but not in `generation.py` -- I've updated it [here](https://github.com/golsun/DialogRPT/blob/master/src/generation.py#L148), please take a look.
* number of candidates to be scored -- if it's too large, you can split candidates into several batches and send them to DialogRPT, similar to [this](https://github.com/golsun/DialogRPT/blob/master/src/score.py#L24)
* if that still doesn't work, I guess you can use two machines, one just for DialoGPT-large and one for DialogRPT, and use API to communicate with each other
* how many DialogRPT models are you using? I recommend at least `updown` and `human_vs_rand` because `updown` doesn't capture context-response relevance.

**Comment by pablogranolabar** [comment]
Hi @golsun, thanks for the quick response!

The two machine idea makes sense, I think I can do that with relative ease if it comes to that.

For the DialogRPT models I am just using updown. So I should ensemble at least updown + human_vs_rand? This application is for a conversational agent that can rerank dialog based on human scoring of the chatbot responses.

**Comment by golsun** [comment]
yes `human_vs_rand` (together with `updown`)should help in that case.
besides `human_vs_rand`, another low-memory way is to decode response with small `top_k` or `top_p`, this should also help the response to be relevant to context.

**Comment by pablogranolabar** [comment]
Another question @golsun. Since DialogRPT has been integrated with HuggingFace transformers library, what are best practices for modifying transformers for human_vs_rand + updown ensembling? For example, after digging around in the installed transformers library via pip3 I'm not finding ensemble.yml.

For production use should I just ditch the transformers library and use your repo instead?

**Comment by pablogranolabar** [comment]
Hi again @golsun. I'm working on ensembling human_vs_rand with updown per your advice, but I'm unsure of the way to proceed with ensemble.yml. Should human_vs_rand and updown be a part of prior with equal weights? Or should human_vs_rand be prior and with updown conditional? Based on the performance reasons above I'm trying to do this with just a two model ensemble as you suggested.

**Comment by golsun** [comment]
hi, in this case, I guess a simple way without dealing with `ensemble.yml` is
```
hvm = get_model('restore/human_vs_machine.pth')
updown = get_model('restore/updown.pth')
score_hvm =  predict(hvm, cxt, hyps)
score_updown =  predict(updown, cxt, hyps)
score_overall = np.sqrt(score_updown * score_hvm)
```
I used geometric mean for `score_overall`, but you can play with some weighted arithmetic mean.
